### PR TITLE
fix(await-thenable): false positive on types extending AsyncDisposable

### PR DESCRIPTION
**Pull request by art-shen**
I had `typescript/await-thenable` set to `error` in a project that uses bun-types. It
started firing on every `await using` in the codebase — `Bun.Subprocess`, any interface
extending `AsyncDisposable`. TypeScript had no complaints. The only fix I could find was
downgrading the rule to `warn`, which masked the problem.

After @camc314 asked for a regression test and a minimal reproducer, I had to build
something that actually reproduced it from scratch. That forced a more precise diagnosis.

When tsgolint processes a file that isn't covered by any tsconfig `include` pattern, it
goes through `CreateInferredProjectProgram`. That function hardcoded
`Target: ScriptTargetES2022` with no `lib`. ES2022 doesn't include
`lib.esnext.disposable.d.ts`. Without it, `Symbol.asyncDispose` has no unique-symbol
entry — `lateBindMember` never registers `[Symbol.asyncDispose]` on any type and
`GetWellKnownSymbolPropertyOfType` returns nil. The rule fires.

For files matched by a tsconfig `include`, the tsconfig's real `lib` loads. Everything
works. The false positive only hits files going through `CreateInferredProjectProgram`.

Why existing tests didn't catch it: they all run from a single compilation where `lib`
always loads correctly. The bug only surfaces in real projects where files go through
`CreateInferredProjectProgram`.

Reproducer: https://github.com/art-shen/tsgolint-await-thenable-reproducer
Issue: #830

---

## The fix

`FindTsConfigParallel` now records `NearestConfig` — the closest ancestor `tsconfig.json`
even when the file doesn't match its `include` pattern. Files outside any include go into
`UnmatchedFiles[nearestConfig]` instead of all landing in `UnmatchedFiles[""]`.
`CreateInferredProjectProgram` parses that tsconfig and uses its `CompilerOptions`. A file
near a project with `"lib": ["esnext"]` now gets that lib. Files with no tsconfig in any
ancestor still fall back to ES2022 defaults.

I also switched `GetWellKnownSymbolPropertyOfType` from exact-name lookup to
prefix-matching on `\xFE@@<name>` with an `@` boundary — so `"match"` can't collide with
`"matchAll@7"`. This is how ts-api-utils handles the same function. I can't prove a
failure before this change in the test suite — the Go checker assigns stable IDs within a
single compilation.

## typescript-go submodule bump

`shim/project/shim.go` on main references `project.ProjectLoadKind`,
`project.SnapshotFSBuilder`, and `ConfigFileRegistryBuilder` methods that only exist in
the newer typescript-go. Someone regenerated the shim but didn't bump the submodule
pointer — main at `adfd2b7` doesn't build. This PR bumps the submodule to `c9f9d601`
(the commit the shim was generated against). `go.work.sum` updates automatically from
the bump.

## Regression tests

- `TestInferredProgramInheritsLibFromNearestTsconfig`: `await using` on an interface
  extending `AsyncDisposable` must not fire when the nearest tsconfig has `"lib": ["esnext"]`.
- `TestInferredProgramNoNearestTsconfig`: `UnmatchedFiles[""]` with no ancestor tsconfig
  falls back to ES2022 defaults; `await-thenable` still fires on `await 0`.
- `TestFindTsConfigParallel_NearestConfig` / `_NoNearestConfig`: verifies
  `Config`/`NearestConfig` semantics for matched and unmatched files.

### CODE REVIEW

File: internal/utils/ts_api_utils.go [hunk 1]
```diff
@@ -171,8 +173,21 @@ func IsThenableType(
...
+// (e.g. "match" vs "matchAll").
 func GetWellKnownSymbolPropertyOfType(t *checker.Type, name string, typeChecker *checker.Checker) *ast.Symbol {
-	return checker.Checker_getPropertyOfType(typeChecker, t, checker.Checker_getPropertyNameForKnownSymbolName(typeChecker, name))
+	prefix := ast.InternalSymbolNamePrefix + "@" + name
+	for _, prop := range checker.Checker_getPropertiesOfType(typeChecker, t) {
+		if prop.Name == prefix || strings.HasPrefix(prop.Name, prefix+"@") {
+			return prop
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (art-shen) [reply internal/utils/ts_api_utils.go:190]: Added `well_known_symbol_test.go` with a table test that feeds crafted property names into the matching logic directly. It covers the actual mismatch (`\xFE@asyncDispose@42` vs `\xFE@asyncDispose@99`), the `match`/`matchAll` collision boundary, and a few other edge cases.

The integration tests in `await_thenable_test.go` can't reproduce different symbol IDs because a single type checker instance resolves everything through the same path.  The unit test sidesteps that by testing the string matching with controlled inputs.

**Comment by art-shen** [comment]
Lint failure is from main — `toStaticValue` in `no_unnecessary_condition.go` has an unused return value from #809. Didn't touch it in this PR.

**Review by camc314 — CHANGES_REQUESTED** [review]
This PR doesn't appear to fix anything - I checked out your branch, reverted your change to ts_api_utils.go, and ran the await_thenable tests, and they all still passed.

The `well_known_symbol_test.go` doesn't really test anything useful - we really need an integration test that tests the whole process end-to-end.

Please open an issue first with a minimal reproduction so it can be investigated to find a clear cause before the problem is attempted to be fixed

Thanks

**Comment by camc314** [comment]
> The integration tests pass on main too. The test fixture resolves all symbols through one path, so IDs happen to match. The bug only shows up when declarations come from different sources (like @types/bun). The unit test exercises the mismatch directly.

Looks like you've identified that the tests already pass on main as well?

**Comment by art-shen** [comment]
@camc314 yes, I explained the whole situation and covered what's possible in `well_known_symbol_test.go`.

Currently, it's impossible to create a reproducer within the tsgolint unit tests as the architecture of the tests masks the source of the problem.

> The bug only shows up when declarations come from different sources (like @types/bun). The unit test exercises the mismatch directly.

**Comment by camc314** [comment]
@art-shen ideally we need to cover this with a proper test - as I mentioned, can you make an issue for this, as well as a minimal repro, so I can confirm that this at least was broken before, and is working now?

**Comment by art-shen** [comment]
@camc314, sure, I'll do my best to demonstrate the error.  It will probably take a bit of time, but I can demonstrate it with a separate reproducible mini-project.

**Comment by camc314** [comment]
Thanks

**Comment by art-shen** [comment]
- Separate issue for this problem: https://github.com/oxc-project/tsgolint/issues/830
- Reproducer: https://github.com/art-shen/tsgolint-await-thenable-reproducer
- CI (typecheck passes, lint fails): https://github.com/art-shen/tsgolint-await-thenable-reproducer/actions/runs/23492127237

**Comment by art-shen** [comment]
@camc314, it was a bit more tricky than I expected, but I have prepared the reproducer (https://github.com/oxc-project/tsgolint/issues/830)

My initial idea about the origin of the problem was not entirely correct.  It was not just a different source, but the source **not included in the tsconfig** that triggers the issue. 

> When a file is outside tsconfig `include`, tsgolint adds it as an extra root to the TypeScript-Go program. That extra-root binding pass is where things go wrong: `getPropertyNameForKnownSymbolName("asyncDispose")` appears to fall back to the no-ID form (`\xFE@asyncDispose`) when the global Symbol constructor isn't yet accessible, so the property gets registered under that name. Later, `GetWellKnownSymbolPropertyOfType` resolves it with an ID suffix (`\xFE@asyncDispose@<n>`) — the exact-name match fails, and the rule concludes the type doesn't implement `AsyncDisposable`.

**Comment by art-shen** [comment]
@camc314, I've updated the pull request and the PR description.  Verified the fix in the reproducer.  Please check it out.

**Comment by camc314** [comment]
See https://github.com/oxc-project/tsgolint/issues/830#issuecomment-4177817047

There are no plans to start respecting `excludes` as it means some files will skip being type checked/having type aware linting run on them which is undesirable.

Thanks for investigating the issue and making this PR.
